src/Waives.Pipelines/Pipeline.cs

```diff
@@ -52,22 +53,35 @@ public class Pipeline : IObservable<WaivesDocument>
     {
         private readonly IHttpDocumentFactory _documentFactory;
         private IObservable<WaivesDocument> _pipeline = Observable.Empty<WaivesDocument>();
-        private Action _onPipelineCompleted = () => { };
+        private Action _onPipelineCompleted;
         private readonly Action<DocumentError> _onDocumentError;
         private Action<DocumentError> _userErrorAction = err => { };
         private readonly IRateLimiter _rateLimiter;
+        private readonly ILogger _logger;
 
-        internal Pipeline(IHttpDocumentFactory documentFactory, IRateLimiter rateLimiter)
+        internal Pipeline(IHttpDocumentFactory documentFactory, IRateLimiter rateLimiter) : this(documentFactory, rateLimiter, Loggers.NoopLogger)
+        { }
+
+        internal Pipeline(IHttpDocumentFactory documentFactory, IRateLimiter rateLimiter, ILogger logger)
         {
             _documentFactory = documentFactory;
             _rateLimiter = rateLimiter ?? throw new ArgumentNullException(nameof(rateLimiter));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
             _onDocumentError = err =>
             {
-                Console.WriteLine($"An error occurred during processing of {err.Document}. " +
-                                  $"The error was: {err.Exception.GetType().Name} {err.Exception.Message}");
+                _logger.Log(LogLevel.Error,
+                    $"An error occurred during processing of document '{err.Document.SourceId}'. " +
```

Comment:
:ok_hand: `An error occurred during processing of document '{err.Document.SourceId}': '{err.Exception.GetType().Name}' '{err.Exception.Message}'"`